src/plugins/intel_cpu/src/nodes/ngram.cpp

```diff
@@ -0,0 +1,164 @@
+        windowSize = k * windowStride;
+        leftPaddingSize = windowStride * leftPad;
+        rightPaddingSize = windowStride * rightPad;
+    }
+}
+
+void Ngram::initSupportedPrimitiveDescriptors() {
+    if (!supportedPrimitiveDescriptors.empty())
+        return;
+
+    addSupportedPrimDesc({{LayoutType::ncsp, InferenceEngine::Precision::FP32},
+                          {LayoutType::ncsp, InferenceEngine::Precision::I32}},
+                         {{LayoutType::ncsp, InferenceEngine::Precision::FP32}},
+                         ref_any,
+                         isDynamicNode());
+}
+
+void Ngram::prepareParams() {
+    const auto& srcDataDims = getParentEdgeAt(0)->getMemoryPtr()->getStaticDims();
+    const auto& srcIndicesDims = getParentEdgeAt(1)->getMemoryPtr()->getStaticDims();
+
+    idcesShapeSize = std::accumulate(srcIndicesDims.begin(), srcIndicesDims.end(), 1, std::multiplies<size_t>());
+    idcesStride = getParentEdgeAt(1)->getMemoryPtr()->GetDescWithType<BlockedMemoryDesc>()->getStrides()[0];
+    numIdces = srcIndicesDims[0];
+
+    windowStride = srcDataDims[1];
+    windowSize = k * windowStride;
+    leftPaddingSize = windowStride * leftPad;
+    rightPaddingSize = windowStride * rightPad;
+}
+
+void Ngram::execute(dnnl::stream strm) {
+    auto* srcData = reinterpret_cast<const float*>(getParentEdgeAt(0)->getMemoryPtr()->GetPtr());
+    auto* srcIndices = reinterpret_cast<const int*>(getParentEdgeAt(1)->getMemoryPtr()->GetPtr());
+    auto* dstData = reinterpret_cast<float*>(getChildEdgeAt(0)->getMemoryPtr()->GetPtr());
+
+    std::vector<size_t> batchLenghts{0};
+    batchLenghts.reserve(numIdces + 1);
+    for (size_t i = idcesStride; i < idcesShapeSize; i += idcesStride) {
+        if (srcIndices[i - idcesStride] != srcIndices[i]) {
+            batchLenghts.push_back(i / idcesStride);
+        }
+    }
+    batchLenghts.push_back(idcesShapeSize / idcesStride);
+
+    /* The following procedure applied to each batch:
+       1. Pad both corners of current embedding with zeros. Left/Right pad are computed depending on k.
+       2. Apply sliding window of windowSize with a step windowStride and form k new embedding vectors for the embedding
+    */
+    parallel_for(batchLenghts.size() - 1, [&](const size_t batchIdx) {
+        size_t srcWindowBias = 0;
+        size_t dstWindowBias = 0;
+
+        const size_t niter = batchLenghts[batchIdx + 1] - batchLenghts[batchIdx];
+        const size_t srcBatchBias = batchLenghts[batchIdx] * windowStride;
+        const size_t dstBatchBias = batchLenghts[batchIdx] * windowStride * k;
+        for (size_t i = 0; i < niter; ++i) {
+            const size_t curLeftPad = leftPad >= i ? leftPaddingSize - i * windowStride : 0;
+            const size_t curRightPad = rightPad >= niter - 1 - i ? rightPaddingSize - (niter - 1 - i) * windowStride : 0;
+
+            memset(dstData + dstBatchBias + dstWindowBias, 0, curLeftPad * sizeof(float));
```

Comment:
Are the curLeftPad and curRightPad comparable with destination blob size? Perhaps it is cheaper to fill all blob by zero once before parallel loop?

Reply:
The ratio of the number of zero and non-zero elements depends on the `k` attribute and the input shape. However, quite often a significant part is occupied by paddings (up to 50-60%). So I changed the logic: all blob is filled by zero before parallel loop,